lib/services/access_request.go

```diff
@@ -113,29 +122,406 @@ type DynamicAccessOracle interface {
+	// once we hit our first approval/denial condition, this review represents the
+	// triggering review for the approval/denial state-transition.
+	var lastReview types.AccessReview
+
+	// Iterate through all reviews and aggregate them against `counts`.
+ProcessReviews:
+	for _, rev := range reviews {
+		lastReview = rev
+		for _, tid := range rev.ThresholdIndexes {
+			idx := int(tid)
+			if len(thresholds) <= idx {
+				return nil, trace.Errorf("threshold index '%d' out of range (this is a bug)", idx)
+			}
+			switch {
+			case rev.ProposedState.IsApproved():
+				counts[idx].approval++
+			case rev.ProposedState.IsDenied():
+				counts[idx].denial++
+			default:
+				return nil, trace.BadParameter("cannot calculate state-transition, unexpected proposal: %s", rev.ProposedState)
+			}
+		}
+
+		// If we hit any denial thresholds, short-circuit immediately
+		for i, t := range thresholds {
+
+			if counts[i].denial >= t.Deny && t.Deny != 0 {
+				denied = true
+				break ProcessReviews
+			}
+		}
+
+		// check for roles that can be transitioned to an approved state
+	CheckRoleApprovals:
+		for role, thresholdSets := range req.GetRoleThresholdMapping() {
+			if _, ok := approved[role]; ok {
+				// role was marked approved during a previous iteration
+				continue CheckRoleApprovals
+			}
+
+			// iterate through all theshold sets.  All sets must have at least
+			// one threshold which has hit its approval count in order for the
+			// role to be considered approved.
+		CheckThresholdSets:
+			for _, tset := range thresholdSets.Sets {
+
+				for _, tid := range tset.Indexes {
+					idx := int(tid)
+					if len(thresholds) <= idx {
+						return nil, trace.Errorf("threshold index out of range %s/%d (this is a bug)", role, tid)
+					}
+					t := thresholds[idx]
+
+					if counts[idx].approval >= t.Approve && t.Approve != 0 {
+						// this set contains a threshold which has met its approval condition.
+						// skip to the next set.
+						continue CheckThresholdSets
+					}
+				}
+
+				// no thresholds met for this set. there may be additional roles/thresholds
+				// which did meet their requirements this iteration, but there is no point
+				// processing them unless this set has also hit its requirements.  we therefore
+				// move immediately to processing the next review.
+				continue ProcessReviews
+			}
+
+			// since we skip to the next review as soon as we see a set which has not hit any of its
+			// approval scenarios, we know that if we get to this point the role must be approved.
+			approved[role] = struct{}{}
+		}
+		// If we got here, then we iterated across all roles in the rtm without hitting any that
+		// had not met their approval scenario.  The request has hit an approved state and further
+		// reviews will not be processed.
+		break ProcessReviews
+	}
+
+	switch {
+	case lastReview.ProposedState.IsApproved():
+		if len(approved) != len(req.GetRoleThresholdMapping()) {
```

Comment:
"not equals" comparison is suspect unless it is guaranteed that `len(approved) <= len(req.GetRoleThresholdMapping())` but it's hard to say. Would be safer to say `if len(approved) < len(req.GetRoleThresholdMapping())`.

Reply:
The `approved` mapping's keys are taken from the rtm, so I'm pretty confident in saying that its guaranteed.